redbox-core/redbox/graph/nodes/processes.py

```diff
@@ -250,20 +270,45 @@ def build_tool_pattern(tools=list[StructuredTool]) -> Callable[[RedboxState], di
     def _tool(state: RedboxState) -> dict[str, Any]:
         state_updates: list[dict] = []
 
-        for tool_id, tool_call_dict in state.get("tool_calls", {}).items():
+        tool_calls = state.get("tool_calls", {})
+        if not tool_calls:
+            log.warning("No tool calls found in state")
+            return None
+
+        for tool_id, tool_call_dict in tool_calls.items():
             tool_call = tool_call_dict["tool"]
 
             if not tool_call_dict["called"]:
                 tool = tools_by_name[tool_call["name"]]
 
-                result_state_update = tool.invoke(tool_call["args"])
-                tool_called_state_update = {"tool_calls": {tool_id: {"called": True}}}
+                if tool is None:
+                    log.warning(f"Tool {tool_call['name']} not found")
+                    continue
 
-                state_updates.append(result_state_update | tool_called_state_update)
+                try:
+                    # Deal with InjectedState
+                    args = tool_call["args"].copy()
+                    if has_injected_state(tool):
+                        args["state"] = state
+
+                    log.info(f"Invoking tool {tool_call['name']} with args {args}")
+
+                    # Invoke the tool
+                    result_state_update = tool.invoke(args) or {}
+                    tool_called_state_update = {"tool_calls": {tool_id: {"called": True, "tool": tool_call}}}
+                    state_updates.append(result_state_update | tool_called_state_update)
+
+                except Exception as e:
```

Comment:
you would imagine that there is a langchain runtime exception... but there is not https://python.langchain.com/docs/how_to/tools_error/#tryexcept-tool-call 😆